### PR TITLE
Implement type-safe attributes

### PR DESCRIPTION
**Pull request by Baltoli**
This PR implements a more limited version of the type-safe attribute infrastructure that we use in the K frontend; by doing so we can construct a canonical list of the attributes that we _actually_ use in the backend, and avoid working with string data when we don't need to be.

The core of the change is a new `attribute_set` structure that _accepts_ attributes permissively from the parser, but only allows key-value lookup via a limited set of whitelisted attribute keys. There are a couple of places in the code that need to access the full set of attributes, ignoring the whitelist:
* When pretty-printing via iteration over the set of stored attributes
* When returning the set of attributes to Pyk via the Python bindings

Both of these code paths have comments to make sure they are used cautiously.

The bulk of the changes are 1:1 and can be reviewed pretty easily in the split diff; there's only a small API change in code that consumes attributes in the backend.

There is a small papercut in that the syntax to reference an attribute key is quite verbose (`attribute_set::key::name`); when we upgrade to C++20 in the near future we get access to `using enum attribute_set::key;` which solves this problem and makes the code nicer to read.

I've tested the change upstream against the K integration test suite.

### CODE REVIEW

File: include/kllvm/ast/attribute_set.h [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+   * be present.
+   */
+  std::shared_ptr<KORECompositePattern> const &get(key k) const;
+
+  /**
+   * Look up an attribute with the specified key that has the form:
+   *
+   *   key_name{}("string data")
+   *
+   * and extract the
```
Comment (Scott-Guest) [inline include/kllvm/ast/attribute_set.h:107]: Looks like this comment is unfinished

**Review by Scott-Guest — APPROVED** [review]
LGTM

**Review by theo25 — APPROVED** [review]
This looks good to me. I left a few minor comments but I will approve and leave to you to decide if you want to address them.

File: lib/ast/attribute_set.cpp [hunk 2]
```diff
@@ -0,0 +1,114 @@
+#include <kllvm/ast/AST.h>
+#include <kllvm/ast/attribute_set.h>
+
+#include <optional>
+
+namespace kllvm {
+
+namespace {
+
+std::unordered_map<attribute_set::key, std::string> const &attribute_table() {
```
[the remark this answers is not in the document]
Comment (Baltoli) [reply lib/ast/attribute_set.cpp:10]: These functions are all contained in an [anonymous namespace](https://en.cppreference.com/w/cpp/language/namespace#Unnamed_namespaces), which is just a slightly more idiomatic way of marking functions as having internal linkage in C++.

**Pull request merged**